tool/teleport/common/agentless.go

```diff
@@ -0,0 +1,724 @@
+		CAPins:             clf.CAPins,
+		Clock:              a.clock,
+	}
+
+	if clf.FIPS {
+		registerParams.CipherSuites = defaults.FIPSCipherSuites
+	}
+
+	certs, err := auth.Register(registerParams)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return certs, trace.Wrap(err)
+}
+
+func (a *agentless) getHostCA(ctx context.Context, client auth.ClientI) (hostSSHCA []byte, hostTLSCA []byte, err error) {
+	cas, err := client.GetCertAuthorities(ctx, types.HostCA, false)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	for _, ca := range cas {
+		for _, key := range ca.GetTrustedSSHKeyPairs() {
+			hostSSHCA = append(hostSSHCA, key.PublicKey...)
+		}
+		for _, key := range ca.GetTrustedTLSKeyPairs() {
+			hostTLSCA = append(hostTLSCA, key.Cert...)
+		}
+	}
+
+	return hostSSHCA, hostTLSCA, nil
+}
+
+func (a *agentless) getOpenSSHCA(ctx context.Context, client auth.ClientI) ([]byte, error) {
+	cas, err := client.GetCertAuthorities(ctx, types.OpenSSHCA, false)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var openSSHCA []byte
+	for _, ca := range cas {
+		for _, key := range ca.GetTrustedSSHKeyPairs() {
+			openSSHCA = append(openSSHCA, key.PublicKey...)
+		}
+	}
+
+	return openSSHCA, nil
+}
+
+func (a *agentless) updateKeysAndConfig(ctx context.Context, clf config.CommandLineFlags, newKeys agentlessKeys) error {
+	exists, err := a.existingKeysPresent(clf)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	configuredHostKeys := []string{filepath.Join(clf.OpenSSHKeysPath, teleportKey)}
+	configuredHostCerts := []string{filepath.Join(clf.OpenSSHKeysPath, teleportCert)}
+	if exists {
+		if err := os.RemoveAll(clf.OpenSSHKeysBackupPath); err != nil {
+			return trace.Wrap(err)
+		}
+		if err := os.Rename(clf.OpenSSHKeysPath, clf.OpenSSHKeysBackupPath); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	if err := tryCreateDefaultAgentlesKeysDir(clf.OpenSSHKeysPath); err != nil {
+		return trace.Wrap(err)
+	}
+
+	fmt.Printf("Writing Teleport keys to %s\n", clf.OpenSSHKeysPath)
+	if err := writeKeys(clf.OpenSSHKeysPath, newKeys); err != nil {
+		if a.defaultKeysDir == clf.OpenSSHKeysPath {
+			rmdirErr := os.RemoveAll(a.defaultKeysDir)
+			if rmdirErr != nil {
+				return trace.NewAggregate(err, rmdirErr)
+			}
+		}
+		return trace.Wrap(err)
+	}
+
+	fmt.Println("Updating OpenSSH config")
+	if err := updateSSHDConfig(
+		clf.OpenSSHConfigPath,
+		agentlessSSHDConfigPath,
+		filepath.Join(clf.OpenSSHKeysPath, teleportOpenSSHCA),
+		configuredHostKeys,
+		configuredHostCerts,
+	); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+func (a *agentless) existingKeysPresent(clf config.CommandLineFlags) (bool, error) {
+	_, err := os.Stat(clf.OpenSSHKeysPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, trace.Wrap(err)
+	}
+	return true, nil
+}
+
+// openSSHInitialJoin
+func (a *agentless) openSSHInitialJoin(ctx context.Context, clf config.CommandLineFlags) error {
```

Comment:
I'd suggest some refactoring around this function because I think it mixes two things: initial joining and rotation for a joined node). You can still re-use a bunch of common logic, but you just have to break it up a bit different.

What do you think about something like below? It makes it very easy for the reader to understand what happens during each phase of rotation.

```go
func (a *agentless) init(ctx context.Context, clf config.CommandLineFlags) error {
  // Connect to auth, register this node, and get a certificate bundle back.
  certs, err := a.register()

  // Reconnect to auth with the new identity.
  client, err := a.connectIdentity(certs)

  // Update key and certificates that will be used by OpenSSH on disk.
  err = a.updateKeysAndConfig(certs)

  // Restart OpenSSH.
  err = a.restart()

  // Register this node with Teleport so it can show up in Teleport inventory.
  err = registerServer()

  return nil
}


func (a *agentless) rotate(ctx context.Context, clf config.CommandLineFlags) error {
  // Read existing identity from disk.
  oldCerts, err := readIdentity()

  // Use existing identity to connect to auth.
  client, err := connectIdentity(oldCerts)

  // Request new host certificates.
  certs, err := client.GenerateHostCerts(oldCerts)

  // Update key and certificates that will be used by OpenSSH on disk.
  err = a.updateKeysAndConfig(certs)

  // Restart OpenSSH.
  err = a.restart()

  // Register this node with Teleport so it can show up in Teleport inventory.
  err = registerServer()

  return nil
}
```

I would suggest a similar approach to the rollback function.

Reply:
In `init`, what's the difference between the time you "register this node" on the first line and the time you do it on the last line?